File: lib/wayback.ts
/**
 * Internet Archive integration for privacy-policy persistence.
 *
 * - `lookupLatestWaybackSnapshot` queries archive.org's availability API
 *   for the most recent capture of a URL.
 * - `submitToWaybackSaveNow` triggers a fresh Save Page Now capture and
 *   returns the snapshot URL.
 *
 * Both fail gracefully — an unreachable archive.org must never break the
 * scrape.
 */

import { safeFetch, validateExternalUrl } from "./security";

const WAYBACK_HOSTS = ["archive.org", "web.archive.org", "www.web.archive.org"];

const AVAILABILITY_MAX_BYTES = 64 * 1024;
const AVAILABILITY_TIMEOUT_MS = 8000;
const SAVE_NOW_TIMEOUT_MS = 25_000;
/**
 * The CDX index lists every capture of a URL in one response. Daily-collapsed
 * over five years of App Store captures is a few KB; the cap is generous so a
 * heavily-archived page never truncates mid-list.
 */
const CDX_MAX_BYTES = 1024 * 1024;
const CDX_TIMEOUT_MS = 20_000;
const CDX_ROW_LIMIT = 5000;

/**
 * archive.org answered but refused to serve — rate-limited (429) or a 5xx.
 * Distinct from "no capture" so callers can back off instead of treating a
 * throttled probe as an empty quarter. `retryAfterMs` mirrors the
 * `Retry-After` header when present.
 */
export class WaybackUnavailableError extends Error {
  readonly retryAfterMs: number | null;
  readonly status: number;

  constructor(status: number, retryAfterMs: number | null, endpoint: string) {
    const label =
      status === 429 ? "rate-limited" : `unavailable (HTTP ${status})`;
    const retry =
      retryAfterMs == null
        ? ""
        : ` — retry after ${Math.ceil(retryAfterMs / 1000)}s`;
    super(`archive.org ${label} for ${endpoint}${retry}`);
    this.name = "WaybackUnavailableError";
    this.status = status;
    this.retryAfterMs = retryAfterMs;
  }
}

export function isWaybackUnavailableError(
  error: unknown
): error is WaybackUnavailableError {
  return error instanceof WaybackUnavailableError;
}

/** Parse a `Retry-After` header (delta-seconds or HTTP-date) into ms. */
export function parseRetryAfterMs(raw: string | null): number | null {
  if (!raw) {
    return null;
  }
  const seconds = Number(raw);
  if (Number.isFinite(seconds) && seconds >= 0) {
    return Math.round(seconds * 1000);
  }
  const when = Date.parse(raw);
  if (Number.isFinite(when)) {
    return Math.max(0, when - Date.now());
  }
  return null;
}

/**
 * Throw for responses that mean "archive.org is refusing us right now".
 * Anything else (200, 404, 3xx that safeFetch already followed) is left to
 * the caller's parser, which returns null for payloads without a capture.
 */
function throwIfUnavailable(response: Response, endpoint: string): void {
  if (response.status === 429 || response.status >= 500) {
    throw new WaybackUnavailableError(
      response.status,
      parseRetryAfterMs(response.headers.get("retry-after")),
      endpoint
    );
  }
}

export interface WaybackSnapshot {
  /** Archive capture timestamp in YYYYMMDDhhmmss form, if available. */
  timestamp?: string;
  /** Full URL to the archived page on web.archive.org. */
  url: string;
}

/**
 * Discriminated result for submitToWaybackSaveNow. Callers can check
 * `result.ok` for success or read `result.error` to surface a failure reason.
 */
export type WaybackSaveResult =
  | { ok: true; snapshot: WaybackSnapshot }
  | { ok: false; error: string };

interface WaybackRequestOptions {
  signal?: AbortSignal;
}

/**
 * Query the availability API for the latest snapshot of `targetUrl`. Returns
 * null when archive.org is unreachable, the JSON is malformed, or no
 * snapshot exists.
 */
export async function lookupLatestWaybackSnapshot(
  targetUrl: string,
  options: WaybackRequestOptions = {}
): Promise<WaybackSnapshot | null> {
  if (!targetUrl) {
    return null;
  }

  const endpoint = `https://archive.org/wayback/available?url=${encodeURIComponent(targetUrl)}`;
  try {
    const { body } = await safeFetch(endpoint, {
      allowedHosts: WAYBACK_HOSTS,
      maxBytes: AVAILABILITY_MAX_BYTES,
      timeoutMs: AVAILABILITY_TIMEOUT_MS,
      signal: options.signal,
      redirect: "follow",
      headers: {
        Accept: "application/json",
        "User-Agent": "privacytracker/1.0 (+privacy-history archiver)",
      },
    });

    let parsed: unknown;
    try {
      parsed = JSON.parse(body.toString("utf8"));
    } catch {
      return null;
    }

    const closest = (
      parsed as {
        archived_snapshots?: {
          closest?: { url?: unknown; timestamp?: unknown; available?: unknown };
        };
      }
    )?.archived_snapshots?.closest;

    if (!closest) {
      return null;
    }
    if (closest.available === false) {
      return null;
    }
    if (typeof closest.url !== "string" || !closest.url.startsWith("http")) {
      return null;
    }

    return {
      url: closest.url,
      timestamp:
        typeof closest.timestamp === "string" ? closest.timestamp : undefined,
    };
  } catch (error) {
    if (isAbortError(error)) {
      throw error;
    }
    return null;
  }
}

/**
 * Trigger a fresh Save Page Now capture. Returns the final snapshot URL
 * (after redirect) or the in-flight URL from `Content-Location` when the
 * save is still in progress.
 */
export async function submitToWaybackSaveNow(
  targetUrl: string,
  options: WaybackRequestOptions = {}
): Promise<WaybackSaveResult> {
  if (!targetUrl) {
    return { ok: false, error: "missing target url" };
  }

  // Save Page Now accepts the full URL with scheme. The earlier
  // `.replace(/^https?:\/\//, (m) => m)` was a no-op left over from a
  // half-finished refactor (the callback returned the matched protocol
  // unchanged) — dropping it.
  const endpoint = `https://web.archive.org/save/${targetUrl}`;

  try {
    const validation = validateExternalUrl(endpoint, {
      allowedHosts: WAYBACK_HOSTS,
    });
    if (!validation.ok) {
      return {
        ok: false,
        error: `Save Page Now URL rejected: ${validation.detail ?? validation.error ?? "invalid URL"}`,
      };
    }

    const response = await fetch(endpoint, {
      method: "GET",
      redirect: "manual",
      signal: withTimeoutSignal(SAVE_NOW_TIMEOUT_MS, options.signal),
      headers: {
        Accept: "text/html,application/xhtml+xml",
        "User-Agent": "privacytracker/1.0 (+privacy-history archiver)",
      },
    });
    try {
      await response.body?.cancel();
    } catch {
      // The body is intentionally ignored. Save Page Now often redirects to a
      // full archived page; downloading that page only to learn its URL can
      // trip response-size caps for large App Store pages.
    }

    const snapshot =
      snapshotFromWaybackHeader(response.headers.get("location"), endpoint) ??
      snapshotFromWaybackHeader(
        response.headers.get("content-location"),
        endpoint
      );
    if (snapshot) {
      return { ok: true, snapshot };
    }

    if (response.status === 429) {
      const retryAfter = response.headers.get("retry-after");
      return {
        ok: false,
        error: retryAfter
          ? `Save Page Now rate-limited; retry after ${retryAfter}s`
          : "Save Page Now rate-limited",
      };
    }
    if (response.status >= 400) {
      return {
        ok: false,
        error: `Save Page Now returned HTTP ${response.status}`,
      };
    }

    // 200 with no snapshot URL: Apple served a 429/503 Wayback declined to
    // archive, or the job is still queueing.
    return {
      ok: false,
      error: `Save Page Now returned ${response.status} without a snapshot URL`,
    };
  } catch (error) {
    if (isAbortError(error)) {
      throw error;
    }
    // Timeouts, DNS/TLS failures, or maxBytes cap all land here.
    const message = error instanceof Error ? error.message : String(error);
    return { ok: false, error: message || "save now request failed" };
  }
}

function extractWaybackTimestamp(url: string): string | undefined {
  const match = url.match(/\/web\/(\d{4,14})(?:[a-z_]+)?\//i);
  return match?.[1];
}

function snapshotFromWaybackHeader(
  raw: string | null,
  baseUrl: string
): WaybackSnapshot | null {
  if (!raw) {
    return null;
  }

  let url: URL;
  try {
    url = new URL(raw, baseUrl);
  } catch {
    return null;
  }

  const validation = validateExternalUrl(url.toString(), {
    allowedHosts: WAYBACK_HOSTS,
  });
  if (!(validation.ok && validation.url)) {
    return null;
  }

  const resolved = validation.url;
  const host = resolved.hostname.toLowerCase();
  if (host !== "web.archive.org" && host !== "www.web.archive.org") {
    return null;
  }
  if (!/^\/web\/\d{4,14}(?:[a-z_]+)?\//i.test(resolved.pathname)) {
    return null;
  }

  const snapshotUrl = resolved.toString();
  return {
    url: snapshotUrl,
    timestamp: extractWaybackTimestamp(snapshotUrl),
  };
}

/**
 * Variant of lookupLatestWaybackSnapshot that asks for the snapshot closest
 * to a specific target timestamp. Used by the historical-import flow.
 * No tolerance window — callers compare the returned timestamp themselves.
 */
export async function lookupWaybackSnapshotNear(
  targetUrl: string,
  targetDate: Date,
  options: WaybackRequestOptions = {}
): Promise<WaybackSnapshot | null> {
  if (!targetUrl) {
    return null;
  }

  const timestamp = formatWaybackTimestamp(targetDate);
  const endpoint = `https://archive.org/wayback/available?url=${encodeURIComponent(targetUrl)}&timestamp=${timestamp}`;
  try {
    const { body, response } = await safeFetch(endpoint, {
      allowedHosts: WAYBACK_HOSTS,
      maxBytes: AVAILABILITY_MAX_BYTES,
      timeoutMs: AVAILABILITY_TIMEOUT_MS,
      signal: options.signal,
      redirect: "follow",
      headers: {
        Accept: "application/json",
        "User-Agent": "privacytracker/1.0 (+privacy-history archiver)",
      },
    });
    // A throttled or failing archive must not masquerade as "no capture" —
    // the historical importer would otherwise record an empty quarter and
    // fire Save Page Now on the strength of a 429.
    throwIfUnavailable(response, "availability API");

    let parsed: unknown;
    try {
      parsed = JSON.parse(body.toString("utf8"));
    } catch {
      return null;
    }

    const closest = (
      parsed as {
        archived_snapshots?: {
          closest?: { url?: unknown; timestamp?: unknown; available?: unknown };
        };
      }
    )?.archived_snapshots?.closest;

    if (!closest) {
      return null;
    }
    if (closest.available === false) {
      return null;
    }
    if (typeof closest.url !== "string" || !closest.url.startsWith("http")) {
      return null;
    }

    return {
      url: closest.url,
      timestamp:
        typeof closest.timestamp === "string" ? closest.timestamp : undefined,
    };
  } catch (error) {
    if (isAbortError(error) || isWaybackUnavailableError(error)) {
      throw error;
    }
    return null;
  }
}

/** One capture from the CDX index. */
export interface WaybackCapture {
  /** Epoch-ms of the capture, parsed from `timestamp`. */
  ms: number;
  /** Wayback `YYYYMMDDhhmmss` capture timestamp. */
  timestamp: string;
  /** Replay URL (`https://web.archive.org/web/<ts>/<original>`). */
  url: string;
}

/**
 * List every successful (HTTP 200) capture of `targetUrl` on or after
 * `from`, oldest first, one per day. One request replaces the up-to-seven
 * availability probes per target the importer used to issue, and lets the
 * caller pick the genuinely closest capture for each target locally.
 *
 * Returns `null` when the index is unreachable or its payload is not the
 * expected array-of-arrays — callers fall back to per-target availability
 * probes. Throws {@link WaybackUnavailableError} on 429 / 5xx so a
 * throttled archive is never mistaken for an unarchived page.
 */
export async function listWaybackCaptures(
  targetUrl: string,
  options: WaybackRequestOptions & { from?: Date } = {}
): Promise<WaybackCapture[] | null> {
  if (!targetUrl) {
    return null;
  }
  const params = new URLSearchParams({
    url: targetUrl,
    output: "json",
    fl: "timestamp,statuscode",
    filter: "statuscode:200",
    collapse: "timestamp:8",
    limit: String(CDX_ROW_LIMIT),
  });
  if (options.from) {
    params.set("from", formatWaybackTimestamp(options.from));
  }
  const endpoint = `https://web.archive.org/cdx/search/cdx?${params.toString()}`;
  try {
    const { body, response } = await safeFetch(endpoint, {
      allowedHosts: WAYBACK_HOSTS,
      maxBytes: CDX_MAX_BYTES,
      timeoutMs: CDX_TIMEOUT_MS,
      signal: options.signal,
      redirect: "follow",
      headers: {
        Accept: "application/json",
        "User-Agent": "privacytracker/1.0 (+privacy-history archiver)",
      },
    });
    throwIfUnavailable(response, "CDX index");
    if (response.status !== 200) {
      return null;
    }

    const text = body.toString("utf8").trim();
    // An unarchived URL yields an empty body (not an empty array).
    if (text === "") {
      return [];
    }
    let parsed: unknown;
    try {
      parsed = JSON.parse(text);
    } catch {
      return null;
    }
    if (!Array.isArray(parsed)) {
      return null;
    }

    const captures: WaybackCapture[] = [];
    const seen = new Set<string>();
    for (const row of parsed) {
      if (!Array.isArray(row) || typeof row[0] !== "string") {
        continue;
      }
      const timestamp = row[0];
      // Header row (`["timestamp","statuscode"]`) and anything malformed.
      if (!/^\d{14}$/.test(timestamp) || seen.has(timestamp)) {
        continue;
      }
      const ms = parseWaybackTimestampMs(timestamp);
      if (ms == null) {
        continue;
      }
      seen.add(timestamp);
      captures.push({
        timestamp,
        ms,
        url: `https://web.archive.org/web/${timestamp}/${targetUrl}`,
      });
    }
    captures.sort((a, b) => a.ms - b.ms);
    return captures;
  } catch (error) {
    if (isAbortError(error) || isWaybackUnavailableError(error)) {
      throw error;
    }
    return null;
  }
}

function withTimeoutSignal(
  timeoutMs: number,
  signal?: AbortSignal
): AbortSignal {
  const timeoutSignal = AbortSignal.timeout(timeoutMs);
  return signal ? AbortSignal.any([signal, timeoutSignal]) : timeoutSignal;
}

export function isAbortError(error: unknown): boolean {
  return error instanceof Error && error.name === "AbortError";
}

/** Convert a Date to YYYYMMDD (UTC) for the availability API. */
function formatWaybackTimestamp(date: Date): string {
  const y = date.getUTCFullYear().toString().padStart(4, "0");
  const m = (date.getUTCMonth() + 1).toString().padStart(2, "0");
  const d = date.getUTCDate().toString().padStart(2, "0");
  return `${y}${m}${d}`;
}

/**
 * Parse a Wayback `YYYYMMDDhhmmss` (or any left-anchored prefix) into epoch-ms.
 * Returns null on malformed input. Short forms are padded to midday UTC so
 * the parse doesn't slide a day across timezones.
 */
export function parseWaybackTimestampMs(
  raw: string | undefined
): number | null {
  if (!raw) {
    return null;
  }
  const padded = `${raw}120000000000`.slice(0, 14);
  const match = padded.match(/^(\d{4})(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})$/);
  if (!match) {
    return null;
  }
  const [, y, mo, d, h, mi, s] = match;
  const ms = Date.UTC(
    Number.parseInt(y, 10),
    Number.parseInt(mo, 10) - 1,
    Number.parseInt(d, 10),
    Number.parseInt(h, 10),
    Number.parseInt(mi, 10),
    Number.parseInt(s, 10)
  );
  return Number.isFinite(ms) ? ms : null;
}

File: core/src/scrape/wayback.rs
//! lib/wayback.ts: the archive.org client the historical import drives —
//! the CDX index listing, the availability API, Save Page Now, and the
//! timestamp arithmetic between them. A 429 or 5xx from the index, the
//! availability API or a replay is `Unavailable`, which the import
//! surfaces to its caller; every other failure is a quiet `None`.
use super::js::truthy;
use crate::{
    jsdate::{self, civil_from_days, days_from_civil},
    jsnum::js_to_number,
    jsstr::{js_encode_uri_component, js_form_encode, js_trim},
    outbound::{self, Fetcher, Reply, Request},
};
use regex::Regex;
use serde_json::Value;
use std::{collections::HashSet, sync::OnceLock};
use url::Url;

pub const WAYBACK_HOSTS: &[&str] = &["archive.org", "web.archive.org", "www.web.archive.org"];
pub const USER_AGENT: &str = "privacytracker/1.0 (+privacy-history archiver)";
const AVAILABILITY_MAX_BYTES: usize = 64 * 1024;
const AVAILABILITY_TIMEOUT_MS: u64 = 8000;
const SAVE_NOW_TIMEOUT_MS: u64 = 25_000;
const CDX_MAX_BYTES: usize = 1024 * 1024;
const CDX_TIMEOUT_MS: u64 = 20_000;
const CDX_ROW_LIMIT: usize = 5000;

/// `WaybackUnavailableError`: archive.org throttling or failing.
#[derive(Debug, Clone, PartialEq)]
pub struct Unavailable {
    pub status: u16,
    pub retry_after_ms: Option<i64>,
    pub message: String,
}

impl Unavailable {
    pub fn new(status: u16, retry_after_ms: Option<i64>, endpoint: &str) -> Self {
        let label = if status == 429 {
            "rate-limited".to_string()
        } else {
            format!("unavailable (HTTP {status})")
        };
        let retry = retry_after_ms.map_or(String::new(), |ms| {
            format!(" — retry after {}s", (ms as f64 / 1000.0).ceil() as i64)
        });
        Self {
            status,
            retry_after_ms,
            message: format!("archive.org {label} for {endpoint}{retry}"),
        }
    }
}

/// wayback.ts `parseRetryAfterMs`: `Number(raw)` first — so `"0"` is zero,
/// not junk — then an HTTP date measured from `now`.
pub fn parse_retry_after_ms(raw: Option<&str>, now: i64) -> Option<i64> {
    let raw = raw.filter(|r| !r.is_empty())?;
    let seconds = js_to_number(&Value::String(raw.to_string()));
    if seconds.is_finite() && seconds >= 0.0 {
        return Some((seconds * 1000.0).round() as i64);
    }
    let when = jsdate::parse(raw)?;
    Some((when - now).max(0))
}

fn unavailable_if(reply: &Reply, endpoint: &str, now: i64) -> Result<(), Unavailable> {
    if reply.status == 429 || reply.status >= 500 {
        return Err(Unavailable::new(
            reply.status,
            parse_retry_after_ms(reply.header("retry-after"), now),
            endpoint,
        ));
    }
    Ok(())
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Snapshot {
    pub url: String,
    pub timestamp: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Capture {
    pub ms: i64,
    pub timestamp: String,
    pub url: String,
}

/// `formatWaybackTimestamp`: the UTC date as `yyyymmdd`.
pub fn format_timestamp(ms: i64) -> String {
    let (year, month, day) = civil_from_days(ms.div_euclid(86_400_000));
    format!("{year:0>4}{month:02}{day:02}")
}

/// `Date.UTC(y, mo, d, h, mi, s)`: JavaScript's field overflow rules
/// (a month past December rolls the year, a day past the month's end rolls
/// the month), two-digit years in 1900, and `TimeClip`.
pub(super) fn date_utc(y: i64, mo0: i64, d: i64, h: i64, mi: i64, s: i64) -> Option<i64> {
    let y = if (0..=99).contains(&y) { y + 1900 } else { y };
    let year = y.checked_add(mo0.div_euclid(12))?;
    let month = mo0.rem_euclid(12) + 1;
    let day = days_from_civil(year, month, 1).checked_add(d.checked_sub(1)?)?;
    let t = day
        .checked_mul(86_400_000)?
        .checked_add(h.checked_mul(3_600_000)?)?
        .checked_add(mi.checked_mul(60_000)?)?
        .checked_add(s.checked_mul(1000)?)?;
    (t.abs() <= 8_640_000_000_000_000).then_some(t)
}

/// `parseWaybackTimestampMs`: a 4–14 digit Wayback timestamp, padded to
/// noon, as epoch milliseconds.
pub fn parse_timestamp_ms(raw: Option<&str>) -> Option<i64> {
    let raw = raw.filter(|r| !r.is_empty())?;
    let padded: String = format!("{raw}120000000000").chars().take(14).collect();
    static RE: OnceLock<Regex> = OnceLock::new();
    let re = RE.get_or_init(|| {
        Regex::new(r"^([0-9]{4})([0-9]{2})([0-9]{2})([0-9]{2})([0-9]{2})([0-9]{2})$")
            .expect("static regex")
    });
    let c = re.captures(&padded)?;
    let n = |i: usize| c[i].parse::<i64>().expect("digits");
    date_utc(n(1), n(2) - 1, n(3), n(4), n(5), n(6))
}

/// `extractWaybackTimestamp` / `timestampFromWaybackUrl`.
pub fn extract_timestamp(url: &str) -> Option<String> {
    static RE: OnceLock<Regex> = OnceLock::new();
    let re = RE
        .get_or_init(|| Regex::new(r"(?i)/web/([0-9]{4,14})(?:[a-z_]+)?/").expect("static regex"));
    re.captures(url).map(|c| c[1].to_string())
}

fn request(
    url: String,
    hosts: &[&str],
    max_bytes: usize,
    timeout_ms: u64,
    headers: &[(&str, &str)],
) -> Request {
    let mut request = Request::apple(url, hosts, max_bytes, timeout_ms);
    request.headers = headers
        .iter()
        .map(|(k, v)| (k.to_string(), v.to_string()))
        .collect();
    request
}

/// `listWaybackCaptures(url, { from })`: `None` when the index cannot be
/// used and the import must probe instead, `Some(vec![])` for an empty
/// index. Rows are the CDX JSON form, header row included, filtered to
/// unique fourteen-digit timestamps and sorted.
pub async fn list_captures(
    fetcher: &dyn Fetcher,
    target_url: &str,
    from_ms: Option<i64>,
    now: i64,
) -> Result<Option<Vec<Capture>>, Unavailable> {
    if target_url.is_empty() {
        return Ok(None);
    }
    let mut params = vec![
        ("url", target_url.to_string()),
        ("output", "json".to_string()),
        ("fl", "timestamp,statuscode".to_string()),
        ("filter", "statuscode:200".to_string()),
        ("collapse", "timestamp:8".to_string()),
        ("limit", CDX_ROW_LIMIT.to_string()),
    ];
    if let Some(from) = from_ms {
        params.push(("from", format_timestamp(from)));
    }
    let query = params
        .iter()
        .map(|(k, v)| format!("{k}={}", js_form_encode(v)))
        .collect::<Vec<_>>()
        .join("&");
    let endpoint = format!("https://web.archive.org/cdx/search/cdx?{query}");
    let req = request(
        endpoint,
        WAYBACK_HOSTS,
        CDX_MAX_BYTES,
        CDX_TIMEOUT_MS,
        &[("Accept", "application/json"), ("User-Agent", USER_AGENT)],
    );
    let Ok(reply) = fetcher.fetch(req).await else {
        return Ok(None);
    };
    unavailable_if(&reply, "CDX index", now)?;
    if reply.status != 200 {
        return Ok(None);
    }
    let body = String::from_utf8_lossy(&reply.body);
    let text = js_trim(&body);
    if text.is_empty() {
        return Ok(Some(vec![]));
    }
    let Ok(Value::Array(rows)) = serde_json::from_str::<Value>(text) else {
        return Ok(None);
    };
    let mut captures = vec![];
    let mut seen = HashSet::new();
    for row in rows {
        let Value::Array(cells) = row else {
            continue;
        };
        let Some(Value::String(timestamp)) = cells.first() else {
            continue;
        };
        let well_formed = timestamp.len() == 14 && timestamp.bytes().all(|b| b.is_ascii_digit());
        if !well_formed || seen.contains(timestamp) {
            continue;
        }
        let Some(ms) = parse_timestamp_ms(Some(timestamp)) else {
            continue;
        };
        seen.insert(timestamp.clone());
        captures.push(Capture {
            ms,
            timestamp: timestamp.clone(),
            url: format!("https://web.archive.org/web/{timestamp}/{target_url}"),
        });
    }
    captures.sort_by_key(|c| c.ms);
    Ok(Some(captures))
}

/// `lookupWaybackSnapshotNear`: the availability API's closest capture to
/// a date, or `None` for anything it cannot answer.
pub async fn lookup_near(
    fetcher: &dyn Fetcher,
    target_url: &str,
    target_ms: i64,
    now: i64,
) -> Result<Option<Snapshot>, Unavailable> {
    if target_url.is_empty() {
        return Ok(None);
    }
    let endpoint = format!(
        "https://archive.org/wayback/available?url={}&timestamp={}",
        js_encode_uri_component(target_url),
        format_timestamp(target_ms)
    );
    let req = request(
        endpoint,
        WAYBACK_HOSTS,
        AVAILABILITY_MAX_BYTES,
        AVAILABILITY_TIMEOUT_MS,
        &[("Accept", "application/json"), ("User-Agent", USER_AGENT)],
    );
    let Ok(reply) = fetcher.fetch(req).await else {
        return Ok(None);
    };
    unavailable_if(&reply, "availability API", now)?;
    let Ok(parsed) = serde_json::from_str::<Value>(&String::from_utf8_lossy(&reply.body)) else {
        return Ok(None);
    };
    let closest = &parsed["archived_snapshots"]["closest"];
    if !truthy(closest) || closest["available"] == Value::Bool(false) {
        return Ok(None);
    }
    let Value::String(url) = &closest["url"] else {
        return Ok(None);
    };
    if !url.starts_with("http") {
        return Ok(None);
    }
    Ok(Some(Snapshot {
        url: url.clone(),
        timestamp: closest["timestamp"].as_str().map(str::to_string),
    }))
}

/// `WaybackSaveResult`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SaveResult {
    Saved(Snapshot),
    Failed(String),
}

/// `submitToWaybackSaveNow`: a raw GET with manual redirects whose
/// Location or Content-Location names the capture; the body is never read.
pub async fn save_now(fetcher: &dyn Fetcher, target_url: &str) -> SaveResult {
    if target_url.is_empty() {
        return SaveResult::Failed("missing target url".to_string());
    }
    let endpoint = format!("https://web.archive.org/save/{target_url}");
    if let Err(e) = outbound::validate(&endpoint, WAYBACK_HOSTS, 2048) {
        return SaveResult::Failed(format!("Save Page Now URL rejected: {}", e.detail));
    }
    let mut req = request(
        endpoint.clone(),
        WAYBACK_HOSTS,
        4 * 1024 * 1024,
        SAVE_NOW_TIMEOUT_MS,
        &[
            ("Accept", "text/html,application/xhtml+xml"),
            ("User-Agent", USER_AGENT),
        ],
    );
    req.follow_redirects = false;
    req.read_body = false;
    let reply = match fetcher.fetch(req).await {
        Ok(reply) => reply,
        Err(message) => {
            return SaveResult::Failed(if message.is_empty() {
                "save now request failed".to_string()
            } else {
                message
            })
        }
    };
    if let Some(snapshot) = snapshot_from_header(reply.header("location"), &endpoint)
        .or_else(|| snapshot_from_header(reply.header("content-location"), &endpoint))
    {
        return SaveResult::Saved(snapshot);
    }
    if reply.status == 429 {
        return SaveResult::Failed(
            match reply.header("retry-after").filter(|h| !h.is_empty()) {
                Some(header) => format!("Save Page Now rate-limited; retry after {header}s"),
                None => "Save Page Now rate-limited".to_string(),
            },
        );
    }
    if reply.status >= 400 {
        return SaveResult::Failed(format!("Save Page Now returned HTTP {}", reply.status));
    }
    SaveResult::Failed(format!(
        "Save Page Now returned {} without a snapshot URL",
        reply.status
    ))
}

/// `snapshotFromWaybackHeader`: a Location resolved against the request,
/// accepted only on the replay host with a `/web/<timestamp>/` path.
fn snapshot_from_header(raw: Option<&str>, base: &str) -> Option<Snapshot> {
    let raw = raw.filter(|r| !r.is_empty())?;
    let url = Url::parse(base).ok()?.join(raw).ok()?;
    let resolved = outbound::validate(url.as_str(), WAYBACK_HOSTS, 2048).ok()?;
    let host = resolved.host_str()?.to_lowercase();
    if host != "web.archive.org" && host != "www.web.archive.org" {
        return None;
    }
    static PATH: OnceLock<Regex> = OnceLock::new();
    let path = PATH
        .get_or_init(|| Regex::new(r"(?i)^/web/[0-9]{4,14}(?:[a-z_]+)?/").expect("static regex"));
    if !path.is_match(resolved.path()) {
        return None;
    }
    let snapshot_url = resolved.to_string();
    Some(Snapshot {
        timestamp: extract_timestamp(&snapshot_url),
        url: snapshot_url,
    })
}

#[cfg(test)]
mod tests {
    use super::{format_timestamp, parse_retry_after_ms, parse_timestamp_ms};

    #[test]
    fn timestamps_and_retry_after_follow_node() {
        // Every pair is from node -e.
        assert_eq!(
            parse_timestamp_ms(Some("20210215120000")),
            Some(1_613_390_400_000)
        );
        // "2021" pads to December 00, which rolls back to 30 November.
        assert_eq!(parse_timestamp_ms(Some("2021")), Some(1_638_230_400_000));
        assert_eq!(
            parse_timestamp_ms(Some("20210230")),
            Some(1_614_686_400_000)
        );
        assert_eq!(parse_timestamp_ms(Some("abc")), None);
        assert_eq!(parse_timestamp_ms(None), None);
        assert_eq!(format_timestamp(1_613_390_400_000), "20210215");
        let now = 1_635_768_000_000; // 2021-11-01T12:00:00Z
        assert_eq!(parse_retry_after_ms(Some("120"), now), Some(120_000));
        assert_eq!(parse_retry_after_ms(Some("0"), now), Some(0));
        assert_eq!(parse_retry_after_ms(Some("1.5"), now), Some(1500));
        assert_eq!(
            parse_retry_after_ms(Some("Mon, 01 Nov 2021 12:02:00 GMT"), now),
            Some(120_000)
        );
        assert_eq!(
            parse_retry_after_ms(Some("Mon, 01 Nov 2021 11:00:00 GMT"), now),
            Some(0)
        );
        assert_eq!(parse_retry_after_ms(Some("soon"), now), None);
    }
}
